File: Factorial/Lua/Yonaba/factorial.lua
-- Factorial algorithm implementation
-- See : http://en.wikipedia.org/wiki/Factorial

-- Recursive tail call implementation
local function factorial_recursive(n)
  assert(n >= 0, ('Argument must be positive'))
  if n == 0 then return 1 end
  return n * factorial_recursive(n-1)
end

-- Iterative implementation
local function factorial_iterative(n)
  local fact = 1
  for i = 1, n do fact = fact * i end
  return fact
end

return {
  recursive = factorial_recursive,
  iterative = factorial_iterative
}

File: Factorial/JavaScript/mitogh/factorial.js
/**
 * Factorial recursive way
 * 
 * n! = 1 if n = 0
 * n! = (n-1)! * n if n > 0
 */
var factorial = function(n){
    return (n <= 1) ? n : (n * factorial(n-1));
}

/**
 * Factorial implemented iterative
 * throught n until n >= 1 
 */
var factorial_iterative = function(n){
    var factorial = 1; 
    while(n >= 1){
        factorial *= n--;
    }
    return factorial;
}
